### PR TITLE
Eccodes adopt

### PR DESCRIPTION
**Pull request by marqh**
Enabling the adoption of EcCodes as a replacement for GribAPI

some low level implementation changes required, but they are limited and I believe supportable from the WMO Manual on Codes

some test changes

this includes GRIB capabilities for Python3, which are running in test

perhaps this should be held as a PR until the pre-release is cut, to make deployment easier

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -54,7 +54,7 @@ install:
       conda install --quiet --file minimal-conda-requirements.txt;
     else
       if [[ "$TRAVIS_PYTHON_VERSION" == 3* ]]; then
-        sed -e '/ecmwf_grib/d' -e '/esmpy/d' -e 's/#.\+$//' conda-requirements.txt | xargs conda install --quiet;
+        sed -e '/esmpy/d' -e 's/#.\+$//' conda-requirements.txt | xargs conda install --quiet;
```
Comment (corinnebosley) [inline .travis.yml]: @marqh @bjlittle @pp-mo Can anybody explain to me what difference this makes please?

File: lib/iris/fileformats/grib/_save_rules.py [hunk 2]
```diff
@@ -257,8 +257,18 @@ def latlon_first_last(x_coord, y_coord, grib):
 def dx_dy(x_coord, y_coord, grib):
     x_step = regular_step(x_coord)
     y_step = regular_step(y_coord)
-    gribapi.grib_set(grib, "DxInDegrees", float(abs(x_step)))
-    gribapi.grib_set(grib, "DyInDegrees", float(abs(y_step)))
+    if x_coord.units == 'degrees':
+        gribapi.grib_set(grib, "iDirectionIncrement",
+                         round(1e6 * float(abs(x_step))))
```
Comment (corinnebosley) [inline lib/iris/fileformats/grib/_save_rules.py]: @marqh What is this rounding and multiplication by 1e6 for?

File: lib/iris/tests/integration/format_interop/test_name_grib.py [hunk 3]
```diff
@@ -72,21 +72,18 @@ def check_common(self, name_cube, grib_cube):
...
-        # Check gribapi version, because we currently have a known load/save
-        # problem with gribapi 1v14 (at least).
-        gribapi_ver = gribapi.grib_get_api_version()
-        gribapi_fully_supported_version = \
-            (StrictVersion(gribapi.grib_get_api_version()) <
-             StrictVersion('1.13'))
+
+        # we currently have a known load/save
+        # problem with gribapi version and zero only data, where min == max.
+
```
Comment (corinnebosley) [inline lib/iris/tests/integration/format_interop/test_name_grib.py]: @marqh Is this still the most appropriate position for this comment to live?  It seems out of place now as it doesn't seem to refer to anything which is being tested here.  

I also don't really understand the comment now; which gribapi version has the problem? Is this just a placeholder for a test that will be added later when the load/save problem is addressed?

File: lib/iris/tests/test_grib_save.py [hunk 4]
```diff
@@ -34,7 +34,7 @@
 
 if tests.GRIB_AVAILABLE:
     import gribapi
```
Comment (corinnebosley) [inline lib/iris/tests/test_grib_save.py]: @marqh I don't understand why this import is still here.  Is this not being replaced?

Comment (marqh) [inline lib/iris/tests/test_grib_save.py]: ECCodes provides a module called gribapi, it is a replacement for GRIB_API

see
https://software.ecmwf.int/wiki/display/ECC/GRIB-API+migration
for more information

thus, the import name is unchanged

File: lib/iris/tests/integration/format_interop/test_name_grib.py (hunk 3 above)
Comment (marqh) [inline lib/iris/tests/integration/format_interop/test_name_grib.py]: perhaps the term 'gribapi version' should just be removed from here?

File: lib/iris/fileformats/grib/_save_rules.py (hunk 2 above)
Comment (marqh) [inline lib/iris/fileformats/grib/_save_rules.py]: the 1e6 is a multiplier to provide values in degrees x 10 ^ -6, as per the WMO manual on codes

the `round` ensures proper rounding to convert to an integer

this is to ensure that the `iDirectionIncrement` is set correctly.  this is a coded key, which are preferred by design in the iris implementation, unlike `DyInDegrees` which is computed.  the eccodes implementation has a subtle bug in its implementation, with different behaviour between py2 and py3

we ought to follow up on this with ECMWF, but I have not yet done this

File: .travis.yml (hunk 1 above)
Comment (marqh) [inline .travis.yml]: this is a nasty hack to make sure `ecmwf_grib` is not installed for python3, where it won't build.  This is no longer required as `eccodes` is py3 compatible

File: lib/iris/tests/integration/format_interop/test_name_grib.py [hunk 5]
```diff
@@ -72,21 +72,18 @@ def check_common(self, name_cube, grib_cube):
...
-                    msg = ('NAMEII cube #{}, "{}" has empty data : '
-                           'SKIPPING test for this cube, as save/load will '
-                           'not currently work with gribabi > 1v12.')
-                    warnings.warn(msg.format(i, name_cube.name()))
-                    continue
+            data = name_cube.data
+            if np.min(data) == np.max(data):
+                msg = ('NAMEII cube #{}, "{}" has empty data : '
+                       'SKIPPING test for this cube, as save/load will '
+                       'not currently work with gribabi > 1v12.')
```
Comment (pelson) [inline lib/iris/tests/integration/format_interop/test_name_grib.py]: Needs replacing with eccodes if this still holds.

File: lib/iris/tests/unit/fileformats/grib/load_convert/test_grid_definition_template_4_and_5.py [hunk 6]
```diff
@@ -34,11 +34,11 @@
 
 from iris.coords import DimCoord
 from iris.fileformats.grib._load_convert import \
-    _MDI as MDI, \
+    _MDIs, \
     grid_definition_template_4_and_5
 from iris.tests import mock
 
-
+MDI = _MDIs[1]
```
Comment (pelson) [inline lib/iris/tests/unit/fileformats/grib/load_convert/test_grid_definition_template_4_and_5.py]: I'm not fond of the order dependence of MDI definition.

File: lib/iris/fileformats/grib/_load_convert.py [hunk 7]
```diff
@@ -172,24 +174,24 @@ def _masker(item):
...
         result = ma.masked
-        if value != _MDI and factor != _MDI:
+        if value not in _MDIs and factor not in _MDIs:
             result = _unscale(value, factor)
     return result
 
 
 # Regulations 92.1.4 and 92.1.5.
-_MDI = 2 ** 32 - 1
+_MDIs = [2 ** 31 - 1, 2 ** 32 - 1]
```
Comment (pelson) [inline lib/iris/fileformats/grib/_load_convert.py]: The only real substantive change is this one, as far as I can see. Sorry for making you pull teeth, but are you able to talk me through why there are 2 MDIs?

The GRIB spec talks of:

**92.1.4**: All bits set to “1” for any value indicates that value is missing. This rule shall not apply to
packed data.

**92.1.5**: If applicable, negative values shall be indicated by setting the most significant bit to “1”

So I agree that for unsigned data the MDI is ```2 ** 32 - 1```, and for signed it is ```2 ** 31 - 1```, but ```2 ** 31 - 1``` most certainly isn't an MDI for signed. Treating it in this way cuts out the extra range that unsigned data gives and mis-represents data as MDI when it is not intended to be.

Comment (pelson) [inline lib/iris/fileformats/grib/_load_convert.py]: I'd be more inclined to codify the "is masked" operation as a function than have them as magic numbers in this way - it will help us avoid dtype checking all over the place.

File: lib/iris/fileformats/grib/_save_rules.py (hunk 2 above)
Comment (pelson) [inline lib/iris/fileformats/grib/_save_rules.py]: Yep. Needs raising, and would also hugely benefit from a comment.

File: lib/iris/fileformats/grib/_load_convert.py (hunk 7 above)
Comment (pelson) [inline lib/iris/fileformats/grib/_load_convert.py]: Can I just verify that this change is actually necessary for ecCodes integration? I can certainly imagine a situation where ecCodes now handles dtypes better and the signed/unsigned issue cropping up, but just want to confirm that it has occurred in the transition between gribapi and ecCodes...

**Comment by dkillick** [comment]
I'm :+1: for adopting eccodes over gribapi. But I think we should hold on merging this here until we have finalised how GRIB file interaction in Iris is dealt with (i.e. with reintroducing `iris-grib`).

File: lib/iris/fileformats/grib/_load_convert.py (hunk 7 above)
Comment (marqh) [inline lib/iris/fileformats/grib/_load_convert.py]: yes, i confirm that this is explicitly needed for ECCodes integration.  the handling of missing values has changed.  I believe the new behaviour is more in line with the WMO regulations, where as the coding approach in current master is based on some interpretations made by gribapi which are not preserved in ECCodes

**Comment by marqh** [comment]
> But I think we should hold on merging this here until we have finalised how GRIB file interaction in Iris is dealt with (i.e. with reintroducing iris-grib).

as i stated here
https://groups.google.com/forum/#!topic/scitools-iris-dev/lFw2_CnaEHU
I think that there is significant benefit in banking this work, however the next steps take place.  I think it is a better place to be to have adopted this change and closed off the GRIB - Dask - ECCodes work

**Comment by marqh** [comment]
@pelson 

i have looked more into the missing data handling: I understand your waryness about this implementation and the multiple missing data values.

as I mentioned, some change is required due to behaviour changes in ECCodes.

I have piloted an approach using 
`gribapi.grib_is_missing`
which has the proper links between relevant missing data value and available values, for different data types

you can see this on a branch on my repo
https://github.com/marqh/iris/pull/64

this alternative explicitly records `None` for missing values.  I think this may be a more robust approach.  There is a small work around for arrays of values, which borrows a tiny bit of logic from the current implementation.  This is an edge case, in my view, with very limited scope for use and is not used in any of our integration tests

I'm happy to replace the commits on this PR with commits from that, if you think that the implementation is preferable.  I have included all relevant changes on that branch, so it is ready to go, in my view, if it seems like a good option.

Although this capability offers the opportunity to adapt our save code to use `gribapi.grib_set_missing` I have not made any changes of this kind.  There would be lots of places this would be required and no tests require it so I have left this to 'potential future work'

cheers
mark

**Comment by marqh** [comment]
i've just force updated this PR with my preferred implementation, using `gribapi.grib_is_missing`

you can see the previous branch at
https://github.com/marqh/iris/tree/eccodesAdoptFirstTry
in case it's useful

I think this is a cleaner approach and it is worth discussing it in situ in this PR

**Comment by marqh** [comment]
however, it seems i have upset travis, who is failing to build my PR, how frustrating

i've tried updating the install (useful anyway) to trigger this, but to no avail

any suggestions on how to ask travis very nicely to process this commit would be gratefully received

File: lib/iris/fileformats/grib/_load_convert.py [hunk 8]
```diff
@@ -158,7 +158,9 @@ def _unscale(v, f):
 
     if isinstance(value, Iterable) or isinstance(factor, Iterable):
         def _masker(item):
-            result = ma.masked_equal(item, _MDI)
+            numerical_mdi = 2 ** 32 - 1
```
Comment (marqh) [inline lib/iris/fileformats/grib/_load_convert.py]: add comment here describing the logic process here, and the lack of examples or integration tests at present

File: lib/iris/fileformats/grib/message.py [hunk 9]
```diff
@@ -459,8 +459,12 @@ def _get_key_value(self, key):
             # By default these values are returned as unhelpful strings but
             # we can use int representation to compare against instead.
             res = gribapi.grib_get(self._message_id, key, int)
+            if gribapi.grib_is_missing(self._message_id, key) == 1:
```
Comment (marqh) [inline lib/iris/fileformats/grib/message.py]: get_key function, with inverted order, os check, then get

add 
`# Implementation of Regulations 92.1.4 and 92.1.5 via eccodes.`
to docstring

File: lib/iris/fileformats/grib/_load_convert.py [hunk 10]
```diff
@@ -158,11 +158,21 @@ def _unscale(v, f):
...
+            # This is a small work around for an edge case, which is not
+            # evident in any of our sample GRIB2 messages, where an array
+            # of header elements contains missing values.
+            # iris.fileformats.grib.message returns these as None, but they
+            # are wanted as a numerical masked array, so a temporary mdi
+            # value is used, selected from a legacy implementation of iris,
+            # to construct the masked array. The valure is transient, only in
+            # scope for this function.
+            numerical_mdi = 2 ** 32 - 1
+            item = [numerical_mdi if i is None else i for i in item]
```
Comment (pelson) [inline lib/iris/fileformats/grib/_load_convert.py]: Wouldn't this be better done as:

```
np.ma.masked_array(item, mask=[i is None for i in item])
```
?

File: lib/iris/fileformats/grib/message.py [hunk 11]
```diff
@@ -481,9 +481,21 @@ def get_computed_key(self, key):
...
         """Return coded keys available in this Section."""
         return self._keys
+
+    def _get_value_or_missing(message_id, key):
+        """
+        Return value of header element, or None if value is encoded as missing.
+        Implementation of Regulations 92.1.4 and 92.1.5 via ECCodes.
+
+        """
+        if gribapi.grib_is_missing(self._message_id, key) == 1:
```
Comment (pelson) [inline lib/iris/fileformats/grib/message.py]: The ```== 1``` is superfluous, I believe.

File: lib/iris/fileformats/grib/_load_convert.py (hunk 10 above)
Comment (marqh) [inline lib/iris/fileformats/grib/_load_convert.py]: I believe not.  If this were implemented, then a mixed type underlying array would be created, with floats and None, so NumPy would have to make an object array.  This would be a behaviour change, which I don't think is desired.

File: lib/iris/fileformats/grib/_load_convert.py [hunk 12]
```diff
@@ -177,30 +187,8 @@ def _masker(item):
...
-#       unsigned 32-bit:
-#           min = 0b000...000 = 0
-#           max = 0b111...110 = 2**32 - 2
-#           MDI = 0b111...111 = 2**32 - 1
-#       signed 32-bit:
-#           MDI = 0b111...111 = 2**32 - 1
-#           min = 0b111...110 = -(2**31 - 2)
-#           max = 0b011...111 = 2**31 - 1
+# Use ECCodes gribapi to recognise missing value
+_MDI = None
```
Comment (pelson) [inline lib/iris/fileformats/grib/_load_convert.py]: Can we now remove this altogether?

File: lib/iris/fileformats/grib/message.py [hunk 13]
```diff
@@ -481,9 +481,24 @@ def get_computed_key(self, key):
...
         else:
-            res = gribapi.grib_get(self._message_id, key)
+            res = self._get_value_or_missing(key)
         return res
 
     def keys(self):
         """Return coded keys available in this Section."""
         return self._keys
+
+    def _get_value_or_missing(self, key, use_int=False):
```
Comment (pelson) [inline lib/iris/fileformats/grib/message.py]: 👍 for having factored this out
Rather than a binary toggle, would you mind passing the dtype of the expected result?

File: lib/iris/fileformats/grib/_load_convert.py (hunk 12 above)
Comment (pelson) [inline lib/iris/fileformats/grib/_load_convert.py]: So the answer is - some of the tests assert that a thing is missing, and this is used for that purpose.
The comment for this line doesn't really help in that regard. Would you mind adding something?

**Comment by pp-mo** [comment]
I've now ported all the changes here to iris_grib : https://github.com/SciTools/iris-grib/pull/92

That is, except for some outstanding integration tests which are still in Iris.
Those bits are in #2887

**Comment by pelson** [comment]
Just to say: thanks for doing this @marqh, and thanks to @pp-mo for migrating it over to iris-grib.
👍
